Add margins support for list elements in PFW plugin

KarolDawidziuk opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#5316](https://github.com/ckeditor/ckeditor4/issues/5316): Feature: Add margins support for list elements in PFW plugin when [pasteTools_keepZeroMargins](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-pasteTools_keepZeroMargins) is ON.
```

## What changes did you make?

As lists are unwrapped during the cleaning process and we lost information about margins, I've added a temporary attribute called `cke-list-style-margins` to save information about zero margins. Then margins are applied during creating a new list when the [pasteTools_keepZeroMargins](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-pasteTools_keepZeroMargins) config option is ON.

To prepare a unit test I've used our [getClipboard](https://github.com/ckeditor/ckeditor4/blob/master/dev/pastetools/getclipboard.html) tool and I've take an appropriate html from each browser.

## Which issues does your PR resolve?

Closes #5316.

Comandeer commented on plugins/pastefromword/filter/default.js:
```diff
@@ -893,6 +893,13 @@
+
+				// Preserve keeping list margins zero when pasteTools_keepZeroMargins is ON. #5316
+				if ( keepZeroMargins ) {
+					innermostContainer.attributes.style = firstLevel1Element.attributes[ 'cke-list-style-margins' ];
```
Is that temporary attribute really needed? From my quick test, it seems that `firstLevel1Element` styles contain `margin` styles at this point and can be used directly instead.

Comandeer reviewed (changes requested):
Looks good at the first sight! However, I'm wondering if the introduction of the temporary variable is really needed. I was able to replace it with a direct reference to list item styles and it still worked in the manual test. Could you check it?

KarolDawidziuk commented:
Thanks, @Comandeer for the review. 
I've simplified code and PR is ready for another check.

Comandeer reviewed (changes requested):
Zero margins seem to work correctly but there is some issue with non-zero ones. I've applied the margin to the list and it's rendered in the editor as the margin between each of the list items when in Word is rendered only before and after the list ([sample document](https://github.com/ckeditor/ckeditor4/files/11100305/list-margins.docx)). Could you check if we can fix it in some easy way?

Additionally, it would be good to add some tests for non-zero margins.

KarolDawidziuk commented:
Thanks @Comandeer for the review.

I've added support for non-zero margins for list by default when ACF will be properly set. In addition I've added missing unit and manual test for non-zero ones and PR is ready for another check.

Comandeer reviewed (changes requested):
The logic looks good and margins are preserved. But there are failing tests due to these changes:

<img width="1401" alt="96 failed Paste from Word tests." src="https://user-images.githubusercontent.com/1078728/230419777-0ed3f4d2-7fd4-47ea-a177-b8beefcf6a92.png">

Could you check them? I assume that it would be enough to update fixtures to include the information about list margins.

KarolDawidziuk commented:
@Comandeer, fixtures updated

Comandeer commented:
Rebased on to the latest `master`.

Comandeer reviewed (changes requested):
I'm wondering about changes to horizontal margins (left and right). We already support them via indent plugins and probably shouldn't touch it at all – especially since the current logic incorrectly handles them (see [margin-left.docx](https://github.com/ckeditor/ckeditor4/files/11200911/margin-left.docx)).

We should probably stick only to the vertical ones (top and bottom).

KarolDawidziuk commented:
@Comandeer handling of the horizontal margins in lists images was removed.

Comandeer commented:
Rebased onto the latest `master`.

Comandeer commented on plugins/pastefromword/filter/default.js:
```diff
@@ -1010,6 +1012,35 @@
+
+					var value = CKEDITOR.tools.convertToPx( parsedStyles[ key ] );
+
+					// Preserve keeping zero list margins when pasteTools_keepZeroMargins is ON. #53163
```
```suggestion
					// Preserve keeping zero list margins when pasteTools_keepZeroMargins is ON (#53163).
```

Comandeer commented on tests/plugins/pastefromword/generated/_fixtures/InlineStyles/expected.html:
Looks like most of the fixtures are the same as on `master`, only formatting differs. Could you revert them to their `master`'s versions?

Comandeer reviewed (changes requested):
Looks good, there is one small improvement left.

KarolDawidziuk commented on plugins/pastefromword/filter/default.js:
```diff
@@ -1010,6 +1012,35 @@
 				}, 0 );
 			}
+
+			function getMargins( styles ) {
```
Fixed.

KarolDawidziuk commented:
@Comandeer PR is ready to review.

Comandeer reviewed (approved):
LGTM! I've just added a missing semicolon between zero margins in the `getMargin()` function.
